core/src/main/scala/kafka/tools/ConsoleConsumer.scala

```diff
@@ -124,39 +124,49 @@ object ConsoleConsumer extends Logging {
-    while (messageCount < maxMessages || maxMessages == -1) {
-      val msg: BaseConsumerRecord = try {
-        consumer.receive()
-      } catch {
-        case _: StreamEndException =>
-          trace("Caught StreamEndException because consumer is shutdown, ignore and terminate.")
-          // Consumer is already closed
-          return
-        case _: WakeupException =>
-          trace("Caught WakeupException because consumer is shutdown, ignore and terminate.")
-          // Consumer will be closed
-          return
-        case e: Throwable =>
-          error("Error processing message, terminating consumer process: ", e)
-          // Consumer will be closed
-          return
-      }
-      messageCount += 1
-      try {
-        formatter.writeTo(new ConsumerRecord(msg.topic, msg.partition, msg.offset, msg.timestamp,
-                                             msg.timestampType, 0, 0, 0, msg.key, msg.value, msg.headers), output)
-      } catch {
-        case e: Throwable =>
-          if (skipMessageOnError) {
-            error("Error processing message, skipping this message: ", e)
-          } else {
+    try {
+      while (messageCount < maxMessages || maxMessages == -1) {
+        val msg: BaseConsumerRecord = try {
+          consumer.receive()
+        } catch {
+          case _: StreamEndException =>
+            trace("Caught StreamEndException because consumer is shutdown, ignore and terminate.")
+            // Consumer is already closed
+            return
+          case _: WakeupException =>
+            trace("Caught WakeupException because consumer is shutdown, ignore and terminate.")
+            // Consumer will be closed
+            return
+          case e: Throwable =>
+            error("Error processing message, terminating consumer process: ", e)
             // Consumer will be closed
-            throw e
+            return
+        }
+        messageCount += 1
+        try {
+          formatter.writeTo(new ConsumerRecord(msg.topic, msg.partition, msg.offset, msg.timestamp,
+            msg.timestampType, 0, 0, 0, msg.key, msg.value, msg.headers), output)
+          if (consumer.isInstanceOf[NewShinyConsumer]) {
+            consumer.asInstanceOf[NewShinyConsumer].updateOffsetForPartition(msg)
           }
+        } catch {
+          case e: Throwable =>
+            if (skipMessageOnError) {
+              error("Error processing message, skipping this message: ", e)
+            } else {
+              // Consumer will be closed
+              throw e
+            }
+        }
+
+        if (checkErr(output, formatter)) {
+          // Consumer will be closed
+          return
+        }
       }
-      if (checkErr(output, formatter)) {
-        // Consumer will be closed
-        return
+    } finally {
+      if (consumer.isInstanceOf[NewShinyConsumer]) {
+        consumer.asInstanceOf[NewShinyConsumer].seekToRealPositionsBeforeExit()
```

Comment:
Why don't we just invoke this in `NewShinyConsumer.cleanup`? It is the final call to `close()` which causes the offsets to get committed, so we just need to set the right position prior to invoking it.